From 1bc3b8adb2ef850a0b90542bfac5f4b90f051e92 Mon Sep 17 00:00:00 2001
From: Evan Huus <eapache@gmail.com>
Date: Tue, 25 Nov 2014 20:27:07 -0500
Subject: [PATCH 6/8] dec-dnart: use pinfo-scoped memory for addresses

They may be accessed during the print phase, at which point packet-scope memory
has already been freed.

Bug: 10724
Change-Id: Ifcf5fc0c0857614edf85349b12dfe605abf6fef7
Reviewed-on: https://code.wireshark.org/review/5498
Reviewed-by: Michael Mann <mmann78@netscape.net>
(cherry picked from commit 8ae9b5363ef210813767bb0cf3af2603092f914c)
Reviewed-on: https://code.wireshark.org/review/6427
---
 epan/dissectors/packet-dec-dnart.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/epan/dissectors/packet-dec-dnart.c b/epan/dissectors/packet-dec-dnart.c
index bd8761e..a4f644e 100644
--- a/epan/dissectors/packet-dec-dnart.c
+++ b/epan/dissectors/packet-dec-dnart.c
@@ -350,12 +350,13 @@ dnet_ntoa(const guint8 *data)
 }
 
 static void
-set_dnet_address(address *paddr_src, address *paddr_tgt)
+set_dnet_address(packet_info *pinfo, address *paddr_src, address *paddr_tgt)
 {
     if (paddr_tgt->type != AT_STRINGZ && paddr_src->type == AT_ETHER) {
         char *addr = dnet_ntoa((const guint8 *)paddr_src->data);
         if (addr != NULL)
-            SET_ADDRESS(paddr_tgt, AT_STRINGZ, 1, addr);
+            SET_ADDRESS(paddr_tgt, AT_STRINGZ, 1,
+                    wmem_strdup(pinfo->pool, addr));
     }
 }
 
@@ -376,10 +377,10 @@ dissect_dec_rt(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree)
     col_set_str(pinfo->cinfo, COL_PROTOCOL, "DEC DNA");
     col_clear(pinfo->cinfo, COL_INFO);
 
-    set_dnet_address(&pinfo->dl_src, &pinfo->net_src);
-    set_dnet_address(&pinfo->dl_src, &pinfo->src);
-    set_dnet_address(&pinfo->dl_dst, &pinfo->net_dst);
-    set_dnet_address(&pinfo->dl_dst, &pinfo->dst);
+    set_dnet_address(pinfo, &pinfo->dl_src, &pinfo->net_src);
+    set_dnet_address(pinfo, &pinfo->dl_src, &pinfo->src);
+    set_dnet_address(pinfo, &pinfo->dl_dst, &pinfo->net_dst);
+    set_dnet_address(pinfo, &pinfo->dl_dst, &pinfo->dst);
 
     offset += 2;
     msg_flags = tvb_get_guint8(tvb, offset);
-- 
2.1.4

